src/Microsoft.Identity.Web/HttpContextExtensions.cs

```diff
@@ -26,7 +30,11 @@ internal static void StoreTokenUsedToCallWebAPI(this HttpContext httpContext, Jw
         /// <returns><see cref="JwtSecurityToken"/> used to call the web API.</returns>
         internal static JwtSecurityToken? GetTokenUsedToCallWebAPI(this HttpContext httpContext)
         {
-            return httpContext.Items[Constants.JwtSecurityTokenUsedToCallWebApi] as JwtSecurityToken;
+            // lock due to https://docs.microsoft.com/en-us/aspnet/core/performance/performance-best-practices?#do-not-access-httpcontext-from-multiple-threads
+            lock(httpContext)
```

Comment:
I don't see any big harm in doing this, however I also don't think this is needed. as the write only happens once during validation of the incoming token. and reads only happens in the controllers which happens after the token validation. I agree there can be multiple reads happening concurrently.

Reply:
Yes, this is my observation as well. Not sure even if multiple reads would happen concurrently...

Reply:
The reason this is needed is because code outside the library could be accessing httpContext.Items at the same time. Application insights may do this, but they are already locking on the httpContext, so if we lock as well, issues can be avoided. I'd like to leave the lock in place as a defensive measure against external code which may corrupt the http context.

Reply:
> The reason this is needed is because code outside the library could be accessing httpContext.Items at the same time. Application insights may do this, but they are already locking on the httpContext, so if we lock as well, issues can be avoided. I'd like to leave the lock in place as a defensive measure against external code which may corrupt the http context.

This assumes the other code is also locking on the HttpContext instance rather than using some other synchronization method like SimaphoreSlim like you do elsewhere. All multi-threading code needs to have an explicit contract for how that process is made safe. In practice you can prepare for concurrent executions of your own APIs, but not somebody else's. I'd remove the lock here.

We've made a similar point to ApplicationInsights and they've gotten better about pre-fetching fields they need when they have definitive control of the context.

Reply:
Makse sense. I believe what David suggest for locking on httpcontext has been consistently using the httpcontext for locking, the semaphore is used around creation of MSAL objects. Seems reasonable to me that we are using the HttpContext for locking in these cases.

---
In reply to: [600634668](https://github.com/AzureAD/microsoft-identity-web/pull/1082#discussion_r600634668) [](ancestors = 600634668)